torch/distributed/optim/zero_redundancy_optimizer.py

```diff
@@ -0,0 +1,529 @@
+# Copyright (c) Facebook, Inc. and its affiliates. All rights reserved.
+#
+# This source code is licensed under the BSD license found in the
+# LICENSE file in the root directory of this source tree.
+
+from collections import OrderedDict, deque
+import copy
+from itertools import chain
+import logging
+from typing import Any, Callable, Dict, List, Optional, Tuple, Type, Deque
+
+import torch
+import torch.distributed as dist
+from torch.nn import Parameter
+from torch._six import container_abcs
+from torch.optim import Optimizer
+import io
+
+__all__ = ["ZeroRedundancyOptimizer"]
+
+
+# Credits:  classy_vision/generic/distributed_util.py
+def _recursive_copy_to_device(value: Any, non_blocking: bool, device: torch.device) -> Any:
+    """
+    Recursively searches lists, tuples, dicts and copies tensors to device if
+    possible. Non-tensor values are passed as-is in the result.
+
+    .. note:  These are all copies, so if there are two objects that reference
+    the same object, then after this call, there will be two different objects
+    referenced on the device.
+    """
+
+    if isinstance(value, torch.Tensor):
+        return value.to(device, non_blocking=non_blocking)
+
+    if isinstance(value, (list, tuple)):
+        values = [_recursive_copy_to_device(val, non_blocking=non_blocking, device=device) for val in value]
+        return values if isinstance(value, list) else tuple(values)
+
+    if isinstance(value, container_abcs.Mapping):
+        return {
+            key: _recursive_copy_to_device(val, non_blocking=non_blocking, device=device) for key, val in value.items()
+        }
+
+    return value
+
+
+def _broadcast_object(
```

Comment:
Sorry for the late comment, but can we modify this to use the existing `broadcast_object_list` in pytorch native now?

Reply:
yes we can, there are a couple of follow up PRs planned, I removed it at some point because there were some deadlocks on CI but I guess that I was not using it right

Reply:
@blefaudeux Can you tell me more about the details about the `deadlocks on CI`? My training procedure also got stuck intermittently, and the error is not reproducible so getting a hard time solving the problem.

Reply:
Is it safer to use `broadcast_object_list`?

Reply:
It's just that it's part of pytorch now, and less code is better code I think. I'm not sure for the reason for deadlocks, a guess was that I was not using `broadcast_object_list` properly since it's guessing the receiving device by context instead of being explicit, but I'm not 100% sure of that